tm2/pkg/sdk/params/handler_test.go

```diff
@@ -1,58 +1,171 @@
-	tu "github.com/gnolang/gno/tm2/pkg/sdk/testutils"
 )
 
-func TestInvalidMsg(t *testing.T) {
-	t.Parallel()
+type testEnv struct {
+	ctx    sdk.Context
+	store  store.Store
+	keeper Keeper
+}
 
-	h := NewHandler(ParamsKeeper{})
-	res := h.Process(sdk.NewContext(sdk.RunTxModeDeliver, nil, &bft.Header{ChainID: "test-chain"}, nil), tu.NewTestMsg())
-	require.False(t, res.IsOK())
-	require.True(t, strings.Contains(res.Log, "unrecognized params message type"))
+type testable interface {
+	Errorf(format string, args ...interface{})
+	FailNow()
 }
 
-func TestQuery(t *testing.T) {
-	t.Parallel()
+func setupTestEnv(t testable, prefix string) testEnv {
+	var (
+		db           = memdb.NewMemDB()
+		paramsCapKey = store.NewStoreKey("paramsCapKey")
+		ms           = store.NewCommitMultiStore(db)
+	)
 
-	env := setupTestEnv()
-	h := NewHandler(env.keeper)
+	ms.MountStoreWithDB(paramsCapKey, iavl.StoreConstructor, db)
+	require.NoError(t, ms.LoadLatestVersion())
 
-	req := abci.RequestQuery{
-		Path: "params/params_test/foo/bar.string",
-	}
+	keeper := NewParamsKeeper(paramsCapKey, prefix)
 
-	res := h.Query(env.ctx, req)
-	require.Nil(t, res.Error)
-	require.NotNil(t, res)
-	require.Nil(t, res.Data)
+	ctx := sdk.NewContext(
+		sdk.RunTxModeDeliver,
+		ms,
+		&bft.Header{Height: 1, ChainID: "test-chain-id"},
+		log.NewNoopLogger(),
+	).WithConsensusParams(&abci.ConsensusParams{
+		Block: &abci.BlockParams{
+			MaxTxBytes:    1024,
+			MaxDataBytes:  1024 * 100,
+			MaxBlockBytes: 1024 * 100,
+			MaxGas:        10 * 1000 * 1000,
+			TimeIotaMS:    10,
+		},
+		Validator: &abci.ValidatorParams{
+			PubKeyTypeURLs: []string{},
+		},
+	})
 
-	env.keeper.SetString(env.ctx, "foo/bar.string", "baz")
+	return testEnv{
+		ctx:    ctx,
+		store:  ctx.Store(paramsCapKey),
+		keeper: keeper,
+	}
+}
+
+func TestParamsHandler_Process(t *testing.T) {
+	h := NewHandler(Keeper{})
+	res := h.Process(sdk.Context{}, nil)
 
-	res = h.Query(env.ctx, req)
-	require.Nil(t, res.Error)
-	require.NotNil(t, res)
-	require.Equal(t, string(res.Data), `"baz"`)
+	assert.False(t, res.IsOK())
+	assert.True(t, strings.Contains(res.Log, errInvalidMsgType.Error()))
 }
 
-func TestQuerierRouteNotFound(t *testing.T) {
-	t.Parallel()
+func TestParamsHandler_Query(t *testing.T) {
+	t.Run("invalid path", func(t *testing.T) {
+		var (
+			prefix = "params_test"
 
-	env := setupTestEnv()
-	h := NewHandler(env.keeper)
-	req := abci.RequestQuery{
-		Path: "params/notfound",
-		Data: []byte{},
-	}
-	res := h.Query(env.ctx, req)
-	require.Error(t, res.Error)
+			testTable = []struct {
+				name string
+				path string
+			}{
+				{
+					"empty path",
+					"",
+				},
+				{
+					"empty key",
+					fmt.Sprintf("params/%s/", prefix),
+				},
+				{
+					"malformed path, no key",
+					fmt.Sprintf("params/%s", prefix),
+				},
+				{
+					"malformed path, no prefix",
+					"params",
+				},
+				{
+					"malformed path, missing prefix",
+					fmt.Sprintf("params/%d", time.Now().Unix()),
+				},
```

Comment:
it panic with this test and the actual implementation you provide for query
```go
{
		"empty prefix, empty key",
		"params//",
}
```

Reply:
I added this case, but it doesn't panic 👀 

[bf1219b](https://github.com/gnolang/gno/pull/3105/commits/bf1219bba6e42d0735ef3a10259dde6ec849f1c7)

Reply:
My bad then; I probably didn't notice some modifications. In any case, I think it's worth more than nothing to add to this test case. :p